sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -123,8 +123,8 @@ object TypeCoercion {
     // We should cast all relative timestamp/date/string comparison into string comparisons
     // This behaves as a user would expect because timestamp strings sort lexicographically.
     // i.e. TimeStamp(2013-01-01 00:00 ...) < "2014" = true
-    case (StringType, DateType) => Some(StringType)
-    case (DateType, StringType) => Some(StringType)
+    case (StringType, DateType) => Some(DateType)
+    case (DateType, StringType) => Some(DateType)
```

Comment:
Doesn't this mean we always find the common type as date when any arbitrary strings are compared to any dates?

Reply:
Yes, it there any issue in your opinion?